aten/src/ATen/native/BucketizationUtils.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+
+inline void searchsorted_maybe_trim_input_tensors(
+  Tensor& trimmed_input,
+  Tensor& trimmed_boundaries,
+  const Tensor& raw_input,
+  const Tensor& raw_boundaries) {
+
+  bool in_is_contiguous = raw_input.is_contiguous();
+  bool bd_is_contiguous = raw_boundaries.is_contiguous();
+
+  if (!in_is_contiguous) {
+    TORCH_WARN_ONCE("input value tensor is non-contiguous, this will lower the performance due to extra data copy "
+      "when converting non-contiguous tensor to contiguous, please use contiguous input value tensor if possible");
+    trimmed_input = raw_input.contiguous();
+  }
+  if (!bd_is_contiguous) {
+    TORCH_WARN_ONCE("input value tensor is non-contiguous, this will lower the performance due to extra data copy "
+      "when converting non-contiguous tensor to contiguous, please use contiguous input value tensor if possible");
+    trimmed_boundaries = raw_boundaries.contiguous();
+  }
+  if (raw_input.dtype() != raw_boundaries.dtype()) {
+    ScalarType input_stype = raw_input.scalar_type();
+    ScalarType boundaries_stype = raw_boundaries.scalar_type();
+    ScalarType common_stype = c10::promoteTypes(input_stype, boundaries_stype);
```

Comment:
If input is a Scalar, it should not participate in type promotion

Reply:
Rather, the correct thing to do here would be (Int and Floating mean any integral or floating type respectively)
1) Int boundaries + Int scalar -> cast to type of boundaries
2) Int boundaries + Int/Floating scalar -> cast to default scalar type 
3) Float boundaries + Int/Floating scalar -> cast to type of boundaries
This scheme is similar to what is currently done for binary operations. 
If boundaries and values are both tensors then c10::promoteTypes is correctly computing output type

Reply:
Will adopt the type promotion rules that we did for binary operation, checked a little bit, they are exactly the same as what is listed above.